dev/tools/controllerbuilder/template/controller.go

```diff
@@ -175,72 +179,82 @@ func (a *Adapter) Create(ctx context.Context, u *unstructured.Unstructured) erro
 
 	status := &krm.{{.Kind}}Status{}
-	// TODO(user): Please add the proto to krm mapping file under apis/{{.Service}}/{{.Version}}
-	if err := krm.Convert_{{.Kind}}_API_v1_To_KRM_status(created, status); err != nil {
-		return fmt.Errorf("update {{.Kind}} status %w", err)
+	// TODO(user): Please add the Status_FromProto mappers under the same package
+	status := {{.Kind}}Status_FromProto(mapCtx, created)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
 	}
+	// TODO(user): (Optional) Please set status.ObservedState
 	status.ObservedState.CreateTime = ToOpenAPIDateTime(created.GetCreateTime())
 	status.ObservedState.UpdateTime = ToOpenAPIDateTime(created.GetUpdateTime())
 	return setStatus(u, status)
 }
 
 func (a *Adapter) Update(ctx context.Context, u *unstructured.Unstructured) error {
+	log := klog.FromContext(ctx).WithName(ctrlName)
+	log.V(2).Info("updating {{.Kind}}", "name", a.fullyQualifiedName())
+	mapCtx := &direct.MapContext{}
 
-	updateMask := &fieldmaskpb.FieldMask{}
-
-	// TODO(user): Add GCP mutable fields.
+	// TODO(user): (Optional) Add GCP mutable fields.
 	// TODO(kcc): Autogen "func immutable()" for each field
 	// TODO(kcc): autogen updateMastk.path for mutable gcp fields. 
+	updateMask := &fieldmaskpb.FieldMask{}
 	if !reflect.DeepEqual(a.desired.Spec.DisplayName, a.actual.DisplayName) {
 		updateMask.Paths = append(updateMask.Paths, "display_name")
 	}
 
-	resource := &{{.Service}}pb.{{.Kind}}{
-		Name: a.fullyQualifiedName(),
-	}
 	desired := a.desired.DeepCopy()
-	err := krm.Convert_{{.Kind}}_KRM_To_API_v1(desired, resource)
-	if err != nil {
-		return fmt.Errorf("converting {{.Kind}} spec to api: %w", err)
+	resource := {{.Kind}}Spec_ToProto(mapCtx, &desired.Spec)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
 	}
 
-	// TODO(user): write the gcp "UPDATE" operation.
-	req := &{{.Service}}pb.Updat{{.Kind}}Request{}
+	// TODO(user): Complete the gcp "UPDATE" or "PATCH" request with required fields.
+	req := &{{.Service}}pb.Update{{.Kind}}Request{
+		Resource:               resource,
+		Project:                a.ProjectID,
+	}
 	op, err := a.gcpClient.Update{{.Kind}}(ctx, req)
 	if err != nil {
-		return fmt.Errorf("{{.Kind}} %s updating failed: %w", resource.Name, err)
+		return fmt.Errorf("updating {{.Kind}} %s failed: %w", a.fullyQualifiedName(), err)
 	}
 	// TODO(user): Adjust the response, depending on the LRO or not.
 	updated, err := op.Wait(ctx)
 	if err != nil {
-		return fmt.Errorf("{{.Kind}} %s waiting update failed: %w", resource.Name, err)
+		return fmt.Errorf("{{.Kind}} %s waiting update failed: %w", a.fullyQualifiedName(), err)
 	}
+	log.V(2).Info("successfully updated {{.Kind}}", "name", a.fullyQualifiedName())
+
 	status := &krm.{{.Kind}}Status{}
-	if err := krm.Convert_{{.Kind}}_API_v1_To_KRM_status(updated, status); err != nil {
-		return fmt.Errorf("update {{.Kind}} status %w", err)
+	status := {{.Kind}}Status_FromProto(mapCtx, updated)
```

Comment:
This is not really a status-from-proto, because the `updated` maps to the `status.observedState`, not `.status`. 
You can look at this example to see how it should be called https://github.com/GoogleCloudPlatform/k8s-config-connector/blob/master/pkg/controller/direct/cloudbuild/workerpool_controller.go#L196

Reply:
Updated. 

Noted that not all beta resources have status.observedState now, we need to manually add this field for them.